File: cellseer/src/cellseer/plots/dqdv.py
from __future__ import annotations

from typing import Any, Literal

import plotly.graph_objects as go

from ..compute import compute_dqdv_dvdq_per_cycle
from ..io import CellData, CyclerName, load_many
from ..style import (
    base_layout,
    build_shared_grid,
    cell_color,
    cycle_fade_color,
    hex_to_rgba,
    interpolate_onto_grid,
)


Direction = Literal["discharge", "charge"]


def _dqdv_points(cell: CellData, cycles: list[int] | None, direction: Direction) -> list[tuple[int, list[float], list[float]]]:
    computed = compute_dqdv_dvdq_per_cycle(cell.curves, direction=direction)
    out: list[tuple[int, list[float], list[float]]] = []
    allowed = set(cycles) if cycles else None
    for cycle, payload in sorted(computed.items()):
        if allowed is not None and cycle not in allowed:
            continue
        dqdv = payload.get("dqdv", {}) or {}
        v = list(dqdv.get("v", []))
        dqdv_vals = list(dqdv.get("dqdv", []))
        if len(v) >= 2 and len(dqdv_vals) >= 2:
            out.append((cycle, v, dqdv_vals))
    return out


def _voltage_range(per_cell: list[list[tuple[int, list[float], list[float]]]]) -> tuple[float, float] | None:
    all_v: list[float] = []
    for rows in per_cell:
        for _, v, _ in rows:
            all_v.extend(v)
    if not all_v:
        return None
    return max(all_v), min(all_v)


def _build_traces_3d(cells: list[CellData], per_cell: list[list[tuple[int, list[float], list[float]]]]) -> list[go.Scatter3d]:
    traces: list[go.Scatter3d] = []
    for i, (cell, rows) in enumerate(zip(cells, per_cell)):
        base = cell.color or cell_color(cell.cell_id or cell.label or f"dataset-{i+1}", i)
        total = len(rows)
        for j, (cycle, v, dqdv) in enumerate(rows):
            traces.append(
                go.Scatter3d(
                    x=v,
                    y=[cycle] * len(v),
                    z=dqdv,
                    mode="lines",
                    name=f"{cell.label} - Cycle {cycle}",
                    legendgroup=cell.cell_id or cell.label or f"group-{i}",
                    showlegend=j == total - 1,
                    line={"color": cycle_fade_color(base, 1 if total <= 1 else j / (total - 1)), "width": 3},
                    hovertemplate=f"{cell.label}<br>Voltage: %{{x:.2f}} V<br>Cycle: %{{y}}<br>dQ/dV: %{{z:.2f}} Ah/V<extra></extra>",
                )
            )
    return traces


def _find_peak(values: list[float]) -> tuple[int, float]:
    idx = max(range(len(values)), key=lambda i: values[i])
    return idx, values[idx]


def _build_traces_2d(cells: list[CellData], per_cell: list[list[tuple[int, list[float], list[float]]]]) -> list[go.Scatter]:
    """Range view matching the web app's dQ/dV 2D 'range' viewMode.

    Each cycle's dQ/dV (Ah/V) is converted to mAh/V (×1000) and resampled onto a
    shared voltage grid (out-of-range → 0). All cycles are drawn as one faint
    joined envelope; the last cycle is drawn bold with a peak diamond.
    """
    # Shared voltage grid across every cell/cycle (web app: buildSharedGrid).
    all_v = [v for rows in per_cell for (_, vs, _) in rows for v in vs]
    v_grid = build_shared_grid(all_v, 2.5, 4.3, 0.002)

    envelopes: list[go.Scatter] = []
    lines: list[go.Scatter] = []
    peaks: list[go.Scatter] = []
    for i, (cell, rows) in enumerate(zip(cells, per_cell)):
        if not rows:
            continue
        base = cell.color or cell_color(cell.cell_id or cell.label or f"dataset-{i+1}", i)
        faint = hex_to_rgba(base, 0.25)

        # Faint joined envelope of every cycle (nulls separate cycles).
        x_parts: list[float | None] = []
        y_parts: list[float | None] = []
        for idx, (cyc, vs, dqdv) in enumerate(rows):
            y_grid = interpolate_onto_grid(vs, [d * 1000.0 for d in dqdv], v_grid, 0.0)
            if idx > 0:
                x_parts.append(None)
                y_parts.append(None)
            x_parts.extend(v_grid)
            y_parts.extend(y_grid)
        if len(rows) > 1:
            envelopes.append(
                go.Scatter(
                    x=x_parts,
                    y=y_parts,
                    mode="lines",
                    name=cell.label,
                    legendgroup=cell.cell_id or cell.label or f"group-{i}",
                    line={"color": faint, "width": 1.5},
                    connectgaps=False,
                    hovertemplate=f"{cell.label}<br>Voltage: %{{x:.2f}} V<br>dQ/dV: %{{y:.2f}} mAh V⁻¹<extra></extra>",
                )
            )

        # Last cycle bold + peak diamond (both on the shared grid).
        last_cyc, last_v, last_dqdv = rows[-1]
        last_y = interpolate_onto_grid(last_v, [d * 1000.0 for d in last_dqdv], v_grid, 0.0)
        peak_idx, peak_val = _find_peak(last_y)
        peak_v = v_grid[peak_idx]
        lines.append(
            go.Scatter(
                x=v_grid,
                y=last_y,
                mode="lines",
                name=f"{cell.label} · Cycle {last_cyc}",
                legendgroup=cell.cell_id or cell.label or f"group-{i}",
                line={"color": base, "width": 2},
                hovertemplate=f"{cell.label}<br>Voltage: %{{x:.2f}} V<br>dQ/dV: %{{y:.2f}} mAh V⁻¹<extra></extra>",
            )
        )
        peaks.append(
            go.Scatter(
                x=[peak_v],
                y=[peak_val],
                mode="markers",
                name="Peak",
                legendgroup=cell.cell_id or cell.label or f"group-{i}",
                showlegend=False,
                marker={"size": 10, "color": base, "symbol": "diamond", "line": {"width": 1, "color": "#fff"}},
                hovertemplate=f"{cell.label} peak<br>Voltage: %{{x:.2f}} V<br>dQ/dV: %{{y:.2f}} mAh V⁻¹<extra></extra>",
            )
        )
    return envelopes + lines + peaks


def _layout_2d() -> dict[str, Any]:
    """2D layout matching the web app's dQ/dV figure (font 10, #e0e0e0 grid)."""
    font = {"family": "Inter, sans-serif", "size": 10, "color": "#3a3a3c"}
    return {
        "font": font,
        "paper_bgcolor": "white",
        "plot_bgcolor": "white",
        "xaxis": {"title": {"text": "Voltage (V)", "font": font}, "tickfont": font, "gridcolor": "#e0e0e0", "showgrid": True},
        "yaxis": {"title": {"text": "dQ/dV (mAh V⁻¹)", "font": font}, "tickfont": font, "gridcolor": "#e0e0e0", "showgrid": True},
        "legend": {"x": 1.02, "y": 1, "xanchor": "left", "yanchor": "top", "orientation": "v", "font": font},
        "margin": {"t": 36, "r": 8, "b": 48, "l": 40},
        "showlegend": True,
    }


def _scene(v_range: tuple[float, float] | None) -> dict[str, Any]:
    return {
        "xaxis": {"title": {"text": "Voltage (V)"}, "range": list(v_range) if v_range else None},
        "yaxis": {"title": {"text": "Cycle"}},
        "zaxis": {"title": {"text": "dQ/dV (Ah/V)"}},
        "camera": {"eye": {"x": 1.25, "y": 1.25, "z": 1.25}},
    }


def dqdv_plot(
    data: Any,
    *,
    cycles: list[int] | None = None,
    mode: str = "3d",
    labels: list[str] | None = None,
    direction: Direction = "discharge",
    cycler: CyclerName = "auto",
) -> go.Figure:
    """Incremental-capacity (dQ/dV vs V) figure.

    mode="3d" (default) stacks cycles along a third axis; mode="2d" overlays
    cycles with peak markers. Accepts anything load() accepts; pass a list
    for multi-cell comparison.
    """
    cells = load_many(data, labels=labels, cycler=cycler)
    per_cell = [_dqdv_points(cell, cycles, direction) for cell in cells]
    if all(len(rows) == 0 for rows in per_cell):
        raise ValueError(
            "No valid dQ/dV cycles found. Ensure input contains charge/discharge records "
            "with non-zero capacity change and voltage span."
        )

    if mode == "2d":
        fig = go.Figure(data=_build_traces_2d(cells, per_cell))
        fig.update_layout(**_layout_2d())
        return fig

    fig = go.Figure(data=_build_traces_3d(cells, per_cell))
    fig.update_layout(
        scene=_scene(_voltage_range(per_cell)),
        scene_aspectmode="manual",
        scene_aspectratio={"x": 1.4, "y": 1.0, "z": 0.7},
        margin={"l": 20, "r": 20, "t": 60, "b": 20},
        font={"family": "Inter, system-ui, sans-serif", "size": 12, "color": "#3a3a3c"},
        paper_bgcolor="white",
        showlegend=True,
    )
    return fig

File: frontend/src/charts/plotting/dqdv.ts
import { cellColor, cellCycleColor, cycleFadeColor } from '../style';
import type { BuildDqDvOpts, Dataset, DqDv2DViewMode, DqDvFigure } from '../types';

// Categorical palette — distinguishable for up to ~12 cells.
const CATEGORICAL_COLORS = [
  '#1f77b4', '#ff7f0e', '#2ca02c', '#9467bd',
  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22',
  '#17becf', '#aec7e8', '#ffbb78', '#98df8a',
];

function categoricalColor(d: Dataset, idx: number): string {
  if (d.color && !d.color.startsWith('#e') && !d.color.startsWith('#f')) return d.color;
  return CATEGORICAL_COLORS[idx % CATEGORICAL_COLORS.length];
}

function hexToRgba(hex: string, opacity: number): string {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  const a = Math.max(0.3, Math.min(1, opacity));
  return `rgba(${r},${g},${b},${a})`;
}

function findPeak(arr: number[]): { index: number; value: number } {
  let index = 0;
  let value = arr[0] ?? 0;
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] > value) {
      value = arr[i];
      index = i;
    }
  }
  return { index, value };
}

function baseColor(d: Dataset, idx: number): string {
  return d.color ?? cellColor(d.id, idx);
}

/**
 * Resolve which of a cell's cycles to highlight.
 *
 * `cycleIndex` is a position in the dashboard's *global* union of every cell's
 * cycles, but `dataset.cycles` holds only the cycles this cell actually has — so
 * indexing it positionally highlights the wrong curve, or none at all when the
 * index runs off the end, whenever a cell is missing cycles (dQ/dV is often
 * computed only every Nth cycle). Match the requested cycle *number*
 * (`selectedCycle`) exactly; if this cell does not have that cycle, highlight
 * nothing for it. Fall back to the positional index only when no cycle number
 * was supplied (legacy callers / tests).
 */
function resolveCycle(
  dataset: Dataset,
  selectedCycle: number | undefined,
  cycleIndex: number,
): Dataset['cycles'][number] | undefined {
  if (selectedCycle != null) {
    // Exact cycle-number match; if this cell lacks that cycle, highlight nothing.
    return dataset.cycles.find((c) => c.cycle === selectedCycle);
  }
  return dataset.cycles[cycleIndex];
}

function voltageExtent(datasets: Dataset[]): [number, number] | undefined {
  let min = Number.POSITIVE_INFINITY;
  let max = Number.NEGATIVE_INFINITY;
  for (const dataset of datasets) {
    for (const cycle of dataset.cycles) {
      for (const v of cycle.x) {
        if (v < min) min = v;
        if (v > max) max = v;
      }
    }
  }
  if (!Number.isFinite(min) || !Number.isFinite(max)) return undefined;
  return [max, min];
}

function smallFont(): { size: number; family: 'Inter' } {
  return { size: 10, family: 'Inter' };
}

function build3D(datasets: Dataset[], opts: BuildDqDvOpts = {}): DqDvFigure {
  const traces: Plotly.Data[] = [];
  datasets.forEach((dataset, i) => {
    const color = baseColor(dataset, i);
    const total = dataset.cycles.length;
    const active = resolveCycle(dataset, opts.selectedCycle, opts.cycleIndex ?? -1);
    dataset.cycles.forEach((cycle, idx) => {
      const isActive = active != null && cycle.cycle === active.cycle;
      traces.push({
        x: cycle.x,
        y: new Array(cycle.x.length).fill(cycle.cycle),
        z: cycle.y,
        type: 'scatter3d',
        mode: 'lines',
        name: `${dataset.label} - Cycle ${cycle.cycle}`,
        line: {
          color: isActive ? color : cycleFadeColor(color, total <= 1 ? 1 : idx / (total - 1)),
          width: isActive ? 5 : 2,
        },
        opacity: isActive ? 1 : 0.6,
        legendgroup: dataset.id,
        showlegend: idx === total - 1,
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>Cycle: %{y}<br>dQ/dV: %{z:.2f} mAh V⁻¹<extra></extra>`,
      });
    });
  });

  const font = smallFont();
  return {
    data: traces,
    layout: {
      font,
      scene: {
        xaxis: { title: { text: 'Voltage (V)', font }, range: voltageExtent(datasets), gridcolor: '#e0e0e0', tickfont: font },
        yaxis: { title: { text: 'Cycle (absolute)', font }, gridcolor: '#e0e0e0', tickfont: font },
        zaxis: { title: { text: 'dQ/dV (mAh V⁻¹)', font }, gridcolor: '#e0e0e0', tickfont: font },
        camera: { eye: { x: 1.25, y: 1.25, z: 1.25 } },
      },
      legend: {
        orientation: 'v' as const,
        x: 1.02,
        y: 0,
        xanchor: 'left' as const,
        yanchor: 'bottom' as const,
        font,
      },
      margin: { t: 36, r: 120, b: 48, l: 8 },
    },
  };
}

function build2D(datasets: Dataset[], opts: BuildDqDvOpts): DqDvFigure {
  const cycleIndex = opts.cycleIndex ?? 0;
  const viewMode: DqDv2DViewMode = opts.viewMode ?? 'baseline';
  const baselineCycle = opts.baselineCycle;
  const baselineCycleIndex = opts.baselineCycleIndex ?? 1;
  const traces: Plotly.Data[] = [];

  if (viewMode === 'range') {
    const rangeTraces: Plotly.Data[] = [];
    const lines: Plotly.Data[] = [];
    const peaks: Plotly.Data[] = [];
    datasets.forEach((dataset, i) => {
      const color = categoricalColor(dataset, i);
      const xParts: (number | null)[] = [];
      const yParts: (number | null)[] = [];
      dataset.cycles.forEach((cycle, idx) => {
        if (idx > 0) {
          xParts.push(null);
          yParts.push(null);
        }
        xParts.push(...cycle.x);
        yParts.push(...cycle.y);
      });
      rangeTraces.push({
        x: xParts,
        y: yParts,
        type: 'scatter',
        mode: 'lines',
        name: dataset.label,
        legendgroup: dataset.id,
        legendgrouptitle: { text: dataset.label },
        line: { color: hexToRgba(color, 0.25), width: 1.5, ...(dataset.dash ? { dash: dataset.dash } : {}) },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
        showlegend: true,
        connectgaps: false,
      });

      const cycle = resolveCycle(dataset, opts.selectedCycle, cycleIndex);
      if (!cycle) return;
      lines.push({
        x: cycle.x,
        y: cycle.y,
        type: 'scatter',
        mode: 'lines',
        name: `${dataset.label} Cycle ${cycle.cycle}`,
        legendgroup: dataset.id,
        showlegend: false,
        line: { color, width: 2, ...(dataset.dash ? { dash: dataset.dash } : {}) },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
      });

      const { index: peakIndex, value: peakValue } = findPeak(cycle.y);
      peaks.push({
        x: [cycle.x[peakIndex]],
        y: [peakValue],
        type: 'scatter',
        mode: 'markers',
        name: `${dataset.label} peak`,
        legendgroup: dataset.id,
        showlegend: false,
        marker: { size: 10, color, symbol: dataset.symbol ?? 'diamond', line: { width: 1, color: '#fff' } },
        hovertemplate: `${dataset.label} peak<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
      });
    });
    traces.push(...rangeTraces, ...lines, ...peaks);
  } else {
    const baselineLines: Plotly.Data[] = [];
    const lines: Plotly.Data[] = [];
    const peaks: Plotly.Data[] = [];
    datasets.forEach((dataset, i) => {
      const color = baseColor(dataset, i);
      const cycle = resolveCycle(dataset, opts.selectedCycle, cycleIndex);
      if (!cycle) return;
      // Explicit reference cycles (light → dark by cycle order). When the caller
      // passes baselineCycles at all (even []), honour it exactly — an empty list
      // means "no references", so nothing faint is drawn until the user pins one.
      // Only when baselineCycles is omitted do we fall back to the legacy single
      // baselineCycle / baselineCycleIndex.
      const refCycles = opts.baselineCycles !== undefined
        ? opts.baselineCycles.filter((c) => c !== cycle.cycle).sort((a, b) => a - b)
        : null;
      if (refCycles) {
        refCycles.forEach((cyc, j) => {
          const ref = dataset.cycles.find((c) => c.cycle === cyc);
          if (!ref) return;
          const t = refCycles.length > 1 ? 0.18 + 0.42 * (j / (refCycles.length - 1)) : 0.3;
          baselineLines.push({
            x: ref.x,
            y: ref.y,
            type: 'scatter',
            mode: 'lines',
            name: `${dataset.label} Cycle ${ref.cycle}`,
            line: { color: cellCycleColor(color, t), width: 1.5, ...(dataset.dash ? { dash: dataset.dash } : {}) },
            hovertemplate: `${dataset.label} Cycle ${ref.cycle}<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
          });
        });
      } else {
        const baseline = baselineCycle != null
          ? dataset.cycles.find((c) => c.cycle === baselineCycle)
          : dataset.cycles[baselineCycleIndex];
        if (baseline) {
          baselineLines.push({
            x: baseline.x,
            y: baseline.y,
            type: 'scatter',
            mode: 'lines',
            name: `${dataset.label} Cycle ${baseline.cycle}`,
            line: { color: hexToRgba(color, 0.4), width: 1.5, ...(dataset.dash ? { dash: dataset.dash } : {}) },
            hovertemplate: `${dataset.label} Cycle ${baseline.cycle}<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
          });
        }
      }
      const { index: peakIndex, value: peakValue } = findPeak(cycle.y);
      const peakVoltage = cycle.x[peakIndex] ?? 0;
      lines.push({
        x: cycle.x,
        y: cycle.y,
        type: 'scatter',
        mode: 'lines',
        name: dataset.label,
        line: { color, width: 2, ...(dataset.dash ? { dash: dataset.dash } : {}) },
        hovertemplate: `${dataset.label}<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<br>Peak: ${peakVoltage.toFixed(2)} V, ${peakValue.toFixed(2)} mAh V⁻¹<extra></extra>`,
      });
      peaks.push({
        x: [peakVoltage],
        y: [peakValue],
        type: 'scatter',
        mode: 'markers',
        name: `${dataset.label} peak`,
        marker: { size: 10, color, symbol: dataset.symbol ?? 'diamond', line: { width: 1, color: '#fff' } },
        hovertemplate: `${dataset.label} peak<br>Voltage: %{x:.2f} V<br>dQ/dV: %{y:.2f} mAh V⁻¹<extra></extra>`,
      });
    });
    traces.push(...baselineLines, ...lines, ...peaks);
  }

  const font = smallFont();
  return {
    data: traces,
    layout: {
      font,
      xaxis: { title: { text: 'Voltage (V)', font }, tickfont: font, gridcolor: '#e0e0e0', showgrid: true },
      yaxis: { title: { text: 'dQ/dV (mAh V⁻¹)', font }, tickfont: font, gridcolor: '#e0e0e0', showgrid: true },
      legend: { x: 1.02, y: 1, xanchor: 'left' as const, yanchor: 'top' as const, orientation: 'v' as const, font },
      margin: { t: 36, r: 8, b: 48, l: 40 },
      showlegend: true,
    },
  };
}

export function buildDqDvFigure(datasets: Dataset[], opts: BuildDqDvOpts = {}): DqDvFigure {
  const mode = opts.mode ?? '3d';
  return mode === '2d' ? build2D(datasets, opts) : build3D(datasets, opts);
}
